provisioner/aws-manager.js

```diff
@@ -0,0 +1,696 @@
+        }
+      });
+    });
+  });
+
+  return ids;
+};
+
+
+/**
+ * Count the capacity of this workerType that are in the states specified
+ * by `states`.  Doing this uses the Capcity key from the workerType's
+ * types dictionary.  Remember that capacity is the number of tasks
+ * that this instance/request will be able to service.
+ * If specified, `extraSpotRequests` is a dictionary which contains a region
+ * and worker type categorized list of outstanding spot requests
+ */
+AwsManager.prototype.capacityForType = function(workerType, states) {
+  assert(workerType);
+  var that = this;
+  var wName = workerType.workerType;
+  var capacity = 0;
+  if (!states) {
+    states = ['running', 'pending', 'spotReq'];
+  }
+
+  // Find instances in the retrevied state and add them to the capacity
+  // according to their declared capacity
+  this.ec2.regions.forEach(function(region) {
+    var rState = that.getApi(region);
+
+    if (!rState[wName]) {
+      return;
+    }
+    
+    var wState = rState[wName];
+
+    if (states.includes('running')) {
+      wState.running.forEach(function(instance) {
+        capacity += workerType.capacityOfType(instance.InstanceType);
+      });
+    }
+
+    if (states.includes('pending')) {
+      wState.pending.forEach(function(instance) {
+        capacity += workerType.capacityOfType(instance.InstanceType);
+      });
+    }
+
+    if (states.includes('spotReq')) {
+      wState.spotReq.forEach(function(request) {
+        capacity += workerType.capacityOfType(request.LaunchSpecification.InstanceType);
+      });
+    }
+  });
+
+  // Extra spot requests are those which known to the provisioner but aren't
+  // available yet through the API.  We want to make sure that they are counted
+  // in the available capacity so that we don't resubmit requests for them
+  // over and over again
+  Object.keys(this.getInternal()).forEach(function(region) {
+    var wState = that.getInternal(region, workerType.workerType);
+
+    var notInApi = 0;
+
+    if (wState && states.includes('spotReq')) {
+      wState.spotReq.forEach(function(sr) {
+        capacity += workerType.capacityOfType(sr.request.LaunchSpecification.InstanceType);
+        notInApi++;
+      });
+      if (notInApi > 0) {
+        debug('%d instances (not capacity) not showing up in API calls', notInApi);
+      }
+    }
+  });
+
+  return capacity;
+};
+
+
+/**
+ * Because the AWS is eventually consistent, it will sometimes take time for
+ * spot requests to show up in the describeSpotInstanceRequests calls for
+ * AWS state.  We will maintain an internal table of these submitted but
+ * not yet visible spot requests so that we can offset the count of a given
+ * instance type for figuring out running capacity.  If the provisioning
+ * process is restarted before the spot request shows up in the api's
+ * state we will lose track of it until it turns into an instance.
+ */
+AwsManager.prototype.trackNewSpotRequest = function(sr) {
```

Comment:
This is so much more awesome than last time :)

We should doc what an `sr` actually is.. the it what we submitted to EC2, or?

Btw, `reconcileInternalState` is really good name, love it :)